### ISSUE TITLE
Never unpublish a package!

### ISSUE THREAD

**Issue by julien-f** [angular-ui/bootstrap#3992]
It seems that angular-bootstrap has been unpublished which causes one of my application (maybe more) to break and it now fails to install correctly. (See [npm deprecate documentation](https://docs.npmjs.com/cli/unpublish#warning)).

If you want to no longer maintain a package, simply deprecate it ([doc](https://docs.npmjs.com/cli/deprecate)).

**Comment by tysoncadenhead** [comment]
Yep, it broke us too.

**Comment by jasonrose** [comment]
This broke my app too, and updating to using the `angular-ui-bootstrap` dependency now causes my npm install to fail because of the peerDependencies on bootstrap 3.1 and angular 1.3.7; I'm using bootstrap 3.3 and and angular 1.4.  I will be creating a separate issue for these, and a PR once I can figure out where angular-ui-bootstrap's package.json gets created.

**Comment by julien-f** [comment]
Yes, this is a real problem, please republish the package as 0.12.1.

**Comment by kentcdodds** [comment]
Borked... :-( Please don't unpublish a package... Ever...

**Comment by Foxandxss** [comment]
That is completely my bad. I am so sorry.

That package was created by a third party person and it was confusing the people on what package to use.

Let me see how can I fix that for you all.

**Comment by kentcdodds** [comment]
Thanks for being responsive @Foxandxss. We all make mistakes. Looking forward to a fix.

**Comment by kentcdodds** [comment]
By the way, the `peerDependencies` should be very loose. Something like this:

```json
{
  "angular": "^1.2.x || >= 1.4.0-beta.0 || >= 1.5.0-beta.0",
  "bootstrap": "^3"
}
```

For more info on [semver ranges](http://semver.npmjs.com/). Also, I'm not sure what version of angular angular-ui-bootstrap is supposed to support, but this should cover you for a while. Give people the choice what version of stuff to use :-)

**Comment by Foxandxss** [comment]
@kentcdodds will check that when I fix this, but for angular, we can only work with 1.3.x on the later release tho.

**Comment by kentcdodds** [comment]
Updated range per your comment.

**Comment by Foxandxss** [comment]
@julien-f @tysoncadenhead @jasonrose @kentcdodds Please confirm if the new package fixed it for you.

**Comment by Foxandxss** [comment]
Hold on, little issue.

**Comment by Foxandxss** [comment]
Well, I think the only problem is where the "main" points now, but shouldn't be a problem.

Please, report back.

**Comment by jasonrose** [comment]
If that version range will pick up angular 1.4.3, then it will work for me.  I personally don't think that this needs bootstrap as a peer dependency, since angularui only cares about bootstrap css and shouldn't dictate that it comes from npm, but that's just my opinion.

**Comment by Foxandxss** [comment]
@jasonrose that is an option as well, yes. I just looked into what `angular-material` was doing and well, angular is a dependency for sure.

**Comment by jasonrose** [comment]
I prefer the updates to `main` in package.json, since it puts it more inline with other angular npm packages.  This means that I can define my module as
`angular.module('myModule', [require('angular-ui-bootstrap')])`, whereas before I needed to require it separately and put the string 'ui.bootstrap' into my dependencies.

**Comment by jasonrose** [comment]
On the other hand, I don't mind publishing a 0.12.1 as previously suggested with the same exact configuration as before, and starting 0.13.0 with these differences.

**Comment by kentcdodds** [comment]
Here's [a blogpost](https://medium.com/@kentcdodds/how-to-distribute-your-angularjs-module-e04d4dd58ddc) (by yours truly) about what @jasonrose is talking about.

**Comment by Foxandxss** [comment]
So, there is an old package, `angular-bootstrap` which is stuck into 0.12.x which I deleted and now restored back just to "un-broke" your projects. I want to know if that is fixed so we can close this issue.

On the other hand, there is a `angular-ui-bootstrap` package which points to 0.13.0 which has the main field for browserify/webpack and also this peerDependencies.

Can we talk about the first issue here and maybe open another issue about the new package?

**Comment by jasonrose** [comment]
Yes, `angular-bootstrap@0.12.2` works for me.

**Comment by Foxandxss** [comment]
Thanks, I appreciate it. If @julien-f confirms it as well, I close this issue. @jasonrose could you please open an issue with what you think we could do for the new package?

**Comment by kentcdodds** [comment]
So, I updated mine to use `0.12.2` and that didn't fix it for me... One moment please.

**Comment by chad-autry** [comment]
The 0.12.2 is not compatible with prior 0.12.X when using browserify-shim

An snipet of my package.json

```json
"browser": {
    "ui.bootstrap.dropdown": "./node_modules/angular-bootstrap/src/dropdown/dropdown.js",
    "ui.bootstrap.collapse": "./node_modules/angular-bootstrap/src/collapse/collapse.js",
    "ui.bootstrap.transition": "./node_modules/angular-bootstrap/src/transition/transition.js",
    "ct.ui.router.extras.core": "./node_modules/ui-router-extras/release/modular/ct-ui-router-extras.core.js",
    "ct.ui.router.extras.sticky": "./node_modules/ui-router-extras/release/modular/ct-ui-router-extras.sticky.js"
  },
  "browserify-shim": {
    "ui.bootstrap.dropdown": {
      "depends": [
        "angular"
      ],
      "exports": "angular.module('ui.bootstrap.dropdown').name"
    },
    "ui.bootstrap.transition": {
      "depends": [
        "angular"
      ],
      "exports": "angular.module('ui.bootstrap.transition').name"
    },
    "ui.bootstrap.collapse": {
      "depends": [
        "angular",
        "ui.bootstrap.transition"
      ],
      "exports": "angular.module('ui.bootstrap.collapse').name"
    },
```

Since the source folder no longer exists, you can see where it wouldn't work any longer.

**Comment by Foxandxss** [comment]
I see. The package was never created by us, it was created by a third party user so I don't really know what he did with it.

Not sure how can we fix that for you. Ideas?

**Comment by kentcdodds** [comment]
I solved it using angular-ui-bootstrap@0.12.1 and changing the main (using
an alias with Webpack, not sure how you do it with browserify) because the
main in that version is incorrect.

- Kent C. Dodds

(Sent from my mobile device, please forgive typos or brevity)

**Comment by chad-autry** [comment]
Previously the src was published along with the dist. This allowed me to alias & require the source files of only those components I was using. Basically I was using npm install, and then configuring a custom build as part of my app.

I've fixed the breakage by aliasing a single built file, and then requiring it. It isn't a big deal for me now just doing dev stuff. But trimming it down to only the packages used could be desireable with a production app.

I'd suggest publishing the source, along with the built artifacts so users could include only the components required.

Thanks!

**Comment by Foxandxss** [comment]
@chad-autry You can generate a custom build from our webpage, but sadly there is no npm package for that.

To fix the old and deprecated package (the one you are using) I would need to break it again and I don't think I want to do that. I cannot bring `src` back without deleting the package yet again and changing it entirely.

I don't think that is wise.

Again, I apologize for this error of mine, but I hope you can move to our new `angular-ui-bootstrap` package which comes with the latest 0.13.1 version.

I am not sure how the guy who created that package did it. He had no access to this repo but yet he created this `angular-bootstrap` package. Maybe you can create a dummy package for your project where you can cherry pick the files you need.

**Comment by chad-autry** [comment]
@Foxandxss I'm suggesting going forward to include the src folder when publishing to github. Don't need to do it for currently existing versions, but could bump the version to include the src.

So far as how the previous guy created the npm package, npm and github aren't linked. He would have simply checked out the repo using the read access everyone has and then published to npm.

**Comment by chad-autry** [comment]
Addition: Include src and templates when publishing to npm. I wasn't yet using a component that required templates, but if I had I'd have needed to include the templates as well.

**Comment by Foxandxss** [comment]
@chad-autry Well this is a topic for a different discussion and so far I don't think we are going to support that.

**Comment by wesleycho** [comment]
One thing we could do is publish updates to the `angular-bootstrap` package as part of our release process - this would allow us to publish a version with the `src` and `templates` directories if users want to consume them individually, instead of consuming the whole package.

We would need to come up with a good solution for users requiring only the pieces they want though.

**Comment by chad-autry** [comment]
I went ahead and made a new issue requesting the addition of /src and /template to npm https://github.com/angular-ui/bootstrap/issues/4016

I'm under the impression angular-bootstrap is dead, long live angular-ui-bootstrap? No need to publish 2 versions. It just additional release artifacts, just like the multiple built versions.

**Comment by Foxandxss** [comment]
@chad-autry We never published 2 versions. In fact, we never published anything.

Both `angular-bootstrap` (the old one) and `angular-ui-bootstrap` were created by third party users that doesn't belong to this team.

We chose `angular-ui-bootstrap` as the name we would like to publish new releases (because it matches other angular-ui packages).

I deleted `angular-bootstrap` aiming to reduce the confusion on "what package to use" so we would only have `angular-ui-bootstrap`. Since that was a bad idea for current users (check this issue) I had to recreate it (but badly, no /src folder) but at least fixed it to the majority of the users.

So the idea is to let this `angular-bootstrap` die and push all users to use `angular-ui-bootstrap`.

I am closing this issue so we can continue talking in the other new ones.
